arangod/RocksDBEngine/RocksDBGeoIndex.cpp

```diff
@@ -263,6 +420,205 @@ class RDBNearIterator final : public IndexIterator {
   std::unique_ptr<rocksdb::Iterator> _iter;
 };
 
+class RDBCoveringIterator final : public IndexIterator {
+ public:
+  /// @brief Construct an RocksDBGeoIndexIterator based on Ast Conditions
+  RDBCoveringIterator(LogicalCollection* collection, transaction::Methods* trx,
+                      RocksDBGeoIndex const* index, geo::QueryParams&& params)
+      : IndexIterator(collection, trx, ReadOwnWrites::no),
+        // geo index never needs to observe own writes since it cannot be used
+        // for an UPSERT subquery
+        _index(index),
+        _covering(std::move(params)) {
+    RocksDBTransactionMethods* mthds =
+        RocksDBTransactionState::toMethods(trx, _collection->id());
+    _iter = mthds->NewIterator(_index->columnFamily(), {});
+    TRI_ASSERT(_index->columnFamily()->GetID() ==
+               RocksDBColumnFamilyManager::get(
+                   RocksDBColumnFamilyManager::Family::GeoIndex)
+                   ->GetID());
+  }
+
+  char const* typeName() const override {
+    return "geo-index-covering-iterator";
+  }
+
+  /// internal retrieval loop
+  template<typename F>
+  inline bool nextToken(F&& cb, size_t limit) {
+    if (_covering.isDone()) {
+      // we already know that no further results will be returned by the index
+      return false;
+    }
+
+    // We keep going, until either we have reached our limit or we have
+    // scanned all intervals delivered by _covering:
+    while (limit > 0 &&
+           (!_covering.isDone() || _scanningInterval < _scan.size())) {
+      if (!_covering.hasNext()) {
+        performScan();
+      }
+
+      while (limit > 0 && _covering.hasNext()) {
+        if (std::forward<F>(cb)(_covering.getNext())) {
+          limit--;
+        }
+        _covering.next();
+      }
+    }
+    return !_covering.isDone() || _scanningInterval < _scan.size();
+  }
+
+  bool nextDocumentImpl(DocumentCallback const& cb, size_t limit) override {
```

Comment:
I think if you make a separate function and only then pass it to `read` it will be more readable:
I mean these places:
`nextDocumentImpl` ... `read`
`nextImpl` ... `read`

Reply:
You have a certain point here. However, I don't want to touch the logic here unnecessarily. This code is anyway copy/pasted from the corresponding code in `RDBNearIterator`. So let's get this PR done first and do stylistic improvements some time later.